src/kubelet/client/client.go

```diff
@@ -0,0 +1,158 @@
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+	"github.com/newrelic/nri-kubernetes/v2/src/prometheus"
+)
+
+const (
+	healthzPath             = "/healthz"
+	defaultHTTPKubeletPort  = 10255
+	defaultHTTPSKubeletPort = 10250
+	defaultTimeout          = time.Millisecond * 5000
+)
+
+// Client implements a client for Kubelet, capable of retrieving prometheus metrics from a given endpoint.
+type Client struct {
+	// TODO: Use a non-sdk logger
+	logger        log.Logger
+	doer          client.HTTPDoer
+	endpoint      url.URL
+	apiServerHost string
+}
+
+type OptionFunc func(kc *Client) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(kc *Client) error {
+		kc.logger = logger
+		return nil
+	}
+}
+
+// New builds a Client using the given options.
+func New(kc kubernetes.Interface, config config.Mock, inClusterConfig *rest.Config, opts ...OptionFunc) (*Client, error) {
+	c := &Client{
+		logger:        log.New(false, io.Discard),
+		apiServerHost: inClusterConfig.Host,
+	}
+
+	for i, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, fmt.Errorf("applying option #%d: %w", i, err)
+		}
+	}
+
+	conn, err := c.setupConnection(kc, tripperWithBearerToken(inClusterConfig.BearerToken), config)
+
+	if err != nil {
+		return nil, fmt.Errorf("connecting to kubelet: %w", err)
+	}
+
+	c.doer = conn.client
+	c.endpoint = conn.url
+
+	return c, nil
+}
+
+func (c *Client) setupConnection(kc kubernetes.Interface, tripperBearerToken http.RoundTripper, config config.Mock) (*connParams, error) {
+	kubeletPort, err := getKubeletPort(kc, config.NodeName)
+	if err != nil {
+		return nil, fmt.Errorf("getting kubelet port: %w", err)
+	}
+
+	conn, err := c.setupLocalConnection(tripperBearerToken, config.NodeIP, kubeletPort)
+	if err == nil {
+		c.logger.Debugf("connected to Kubelet directly with nodeIP")
+		return conn, nil
+	}
+
+	c.logger.Debugf("Kubelet connection with nodeIP failed: %v", err)
+	c.logger.Debugf("Connecting to kubelet directly with API proxy")
+
+	conn, err = checkConnectionAPIProxy(c.apiServerHost, config.NodeName, tripperBearerToken)
+	if err != nil {
+		return nil, fmt.Errorf("creating connection parameters for API proxy: %w", err)
+	}
+
+	return conn, nil
+}
+
+func (c *Client) setupLocalConnection(tripperWithBearerToken http.RoundTripper, nodeIP string, portInt int32) (*connParams, error) {
+	c.logger.Debugf("connecting to kubelet directly with nodeIP")
+	var err error
+
+	port := fmt.Sprintf("%d", portInt)
+	hostURL := net.JoinHostPort(nodeIP, port)
+
+	var conn *connParams
+
+	switch portInt {
+	case defaultHTTPKubeletPort:
+		if conn, err = checkConnectionHTTP(hostURL); err == nil {
+			return conn, nil
+		}
+
+	case defaultHTTPSKubeletPort:
+		if conn, err = checkConnectionHTTPS(hostURL, tripperWithBearerToken); err == nil {
+			return conn, nil
+		}
```

Comment:
This would mean that users won't be able to force https on the default HTTP port and viceversa.

It might be something strange to do but I'd prefer to not limit ourselves in that regard: If possible, users should be able to override the protocol and skip this autodetection.